src/libfsm/print/ir.h

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2018 Katherine Flavel
+ *
+ * See LICENCE for the full copyright terms.
+ */
+
+#ifndef FSM_INTERNAL_IR_H
+#define FSM_INTERNAL_IR_H
+
+/*
+ * Codegen IR intended for C-like switch statements
+ *
+ * The fsm_state structs don't exist from this IR onwards.
+ * These nodes are always DFA. No epsilons, No duplicate edge labels,
+ * so this IR takes advantage of that.
+ *
+ * But this is not neccessarily a minimal DFA. By this stage,
+ * we're done with graph algorithmics.
+ */
+
+enum ir_strategy {
+	IR_NONE     = 1 << 0,
+	IR_SAME     = 1 << 1,
+	IR_COMPLETE = 1 << 2,
+	IR_PARTIAL  = 1 << 3,
+	IR_DOMINANT = 1 << 4,
+	IR_ERROR    = 1 << 5,
+	IR_TABLE    = 1 << 6
+};
+
+struct ir_range {
+	unsigned char start;
```

Comment:
Is it worth noting that this is inclusive? Is this something that will need to be adapted for UTF-8?

Reply:
These will always be bytes - for UTF-8, the ranges here describe the values for a single byte within a UTF-8 sequence. Whereas for the AST, i think we'll have unicode codepoints there.

Reply:
Oh, so this is after the code points are chunked into bytes. Right.